fvm/src/kernel/default.rs

```diff
@@ -1006,113 +1006,93 @@ impl<C> EventOps for DefaultKernel<C>
+            return Err(syscall_error!(IllegalArgument; "event exceeded max entries: {} > {MAX_NR_ENTRIES}", event_headers.len()).into());
         }
 
         let mut key_offset: usize = 0;
         let mut val_offset: usize = 0;
 
-        let mut entries: Vec<Entry> = Vec::with_capacity(nr_entries);
-        for i in 0..nr_entries {
-            let offset = i * BYTES_PER_ENTRY;
-            let view = &raw_evt[offset..offset + BYTES_PER_ENTRY];
-
-            // parse the fixed sized fields from the raw_evt buffer
-            //
-            let entry_fixed: EntryFixed = unsafe {
-                std::mem::transmute::<[u8; BYTES_PER_ENTRY], EntryFixed>(view.try_into().unwrap())
-            };
-
+        let mut entries: Vec<Entry> = Vec::with_capacity(event_headers.len());
+        for header in event_headers {
             // make sure that the fixed parsed values are within bounds before we do any allocation
-            //
-            let flags = entry_fixed.flags;
+            let flags = header.flags;
             if Flags::from_bits(flags.bits()).is_none() {
                 return Err(
                     syscall_error!(IllegalArgument; "event flags are invalid: {}", flags.bits())
                         .into(),
                 );
             }
-            if entry_fixed.key_len > MAX_KEY_LEN as u32 {
-                let tmp = entry_fixed.key_len;
+            if header.key_len > MAX_KEY_LEN as u32 {
+                let tmp = header.key_len;
                 return Err(syscall_error!(IllegalArgument; "event key exceeded max size: {} > {MAX_KEY_LEN}", tmp).into());
             }
-            if entry_fixed.val_len > MAX_VALUE_LEN as u32 {
-                let tmp = entry_fixed.val_len;
+            if header.val_len > MAX_VALUE_LEN as u32 {
+                let tmp = header.val_len;
                 return Err(syscall_error!(IllegalArgument; "event value exceeded max size: {} > {MAX_VALUE_LEN}", tmp).into());
             }
-            if entry_fixed.codec != IPLD_RAW {
-                let tmp = entry_fixed.codec;
+            if header.codec != IPLD_RAW {
+                let tmp = header.codec;
                 return Err(
                     syscall_error!(IllegalCodec; "event codec must be IPLD_RAW, was: {}", tmp)
                         .into(),
                 );
             }
 
             // parse the variable sized fields from the raw_key/raw_val buffers
-            //
-            let key = match std::str::from_utf8(
-                &raw_key[key_offset..key_offset + entry_fixed.key_len as usize],
-            ) {
-                Ok(s) => s,
-                Err(e) => {
-                    return Err(syscall_error!(IllegalArgument, "invalid event key: {}", e).into())
-                }
-            };
-            let value = &raw_val[val_offset..val_offset + entry_fixed.val_len as usize];
+            let key = &event_keys
+                .get(key_offset..key_offset + header.key_len as usize)
```

Comment:
bounds checking.

No need to check for overflows as we're limiting the sizes of the key/value lengths.